poll: Support read-only poll widget UI.

PIG208 opened the pull request:
The UI follows the webapp until we get a new design.

The dark theme colors were tentatively picked. The `TextStyle`s are the same for both light and dark theme. All the styling are based on values taken from the webapp.

References:

  - light theme:

    https://github.com/zulip/zulip/blob/2011e0df760cea52c31914e7b77d9b4e38e9ee74/web/styles/widgets.css#L138-L185 https://github.com/zulip/zulip/blob/2011e0df760cea52c31914e7b77d9b4e38e9ee74/web/styles/dark_theme.css#L358

  - dark theme:

    https://github.com/zulip/zulip/blob/2011e0df760cea52c31914e7b77d9b4e38e9ee74/web/styles/dark_theme.css#L966-L987

Fixes #165.

PIG208 commented:
| light | dark|
| - | - |
| ![Screenshot from 2024-08-22 22-30-32](https://github.com/user-attachments/assets/9db663bf-b8a1-4c76-83b8-85b22cf3c729) | ![Screenshot from 2024-08-22 22-31-01](https://github.com/user-attachments/assets/41d1d050-4078-4afb-bfd1-8c9891928eb1) |

chrisbobbe commented on lib/widgets/poll.dart:
```diff
@@ -0,0 +1,77 @@
+                      child: Text(option.text,
+                        style: textStylePollBold.copyWith(fontSize: 16))),
+                    if (option.voters.isNotEmpty)
+                      Text('(${getOptionVoterNames(option)})', style: textStylePollNames),
```
I think this might need a TODO for localizing the parenthesis characters (like to use full-width parens in Chinese).

chrisbobbe commented on lib/widgets/poll.dart:
```diff
@@ -0,0 +1,77 @@
+
+    String getOptionVoterNames(PollOption option) => option.voters.map((userId) =>
+      store.users[userId]?.fullName
+        ?? zulipLocalizations.unknownUserName).join(', ');
```
Let's add a `TODO(i18n)` like this:

```dart
        // TODO(i18n): List formatting, like you can do in JavaScript:
        //   new Intl.ListFormat('ja').format(['Chris', 'Greg', 'Alya', 'Zixuan'])
        //   // 'Chris、Greg、Alya、Zixuan'
```

(we have this TODO in a few places already)

chrisbobbe commented on lib/widgets/poll.dart:
```diff
@@ -0,0 +1,77 @@
+
+    Text question = (poll.question.isNotEmpty)
+      ? Text(poll.question, style: textStylePollBold)
+      : Text(zulipLocalizations.pollWidgetQuestionMissing,
+        style: textStylePollBold.copyWith(fontStyle: FontStyle.italic));
```
```suggestion
      : Text(zulipLocalizations.pollWidgetQuestionMissing,
          style: textStylePollBold.copyWith(fontStyle: FontStyle.italic));
```

chrisbobbe reviewed (commented):
Thanks! Small comments below.

PIG208 commented:
Thanks! Updated the PR.

chrisbobbe commented:
Thanks! LGTM except bump on https://github.com/zulip/zulip-flutter/pull/912#discussion_r1737264933 , and I'll mark this for Greg's review.

PIG208 commented on lib/model/content.dart:
```diff
@@ -74,6 +75,12 @@ mixin UnimplementedNode on ContentNode {
 
 sealed class ZulipMessageContent {}
 
+class PollContent implements ZulipMessageContent {
```
While this is pretty short, I'm not sure if we need a better way to organize `ZulipMessageContent`, `PollContent`, and `MessageContent`. The file `lib/model/content.dart` is predominantly about our AST for Zulip HTML content.

PIG208 commented:
Rebased the PR.

PIG208 commented:
Pushed some simplifications to `test/widgets/poll_test.dart`.

PIG208 commented:
Update: use `spacing` when applicable and fix an issue with large vote numbers:

| before | after |
| - | - |
| ![Screenshot from 2024-09-13 22-28-26](https://github.com/user-attachments/assets/3f52942e-3b6c-423a-84df-981c12b697b8) | ![Screenshot from 2024-09-13 22-28-56](https://github.com/user-attachments/assets/40b3f387-f995-4af6-831e-b416f5ad1fe9) |

PIG208 commented on lib/widgets/poll.dart:
```diff
@@ -0,0 +1,85 @@
+      return option.voters
+        .map((userId) =>
+          store.users[userId]?.fullName ?? zulipLocalizations.unknownUserName)
+        .join(', ');
```
Indented this in a style taken from `ListNode.debugDescribeChildren`, to avoid going over 80 characters.
```dart
  @override
  List<DiagnosticsNode> debugDescribeChildren() {
    return items
      .mapIndexed((i, nodes) =>
        _BlockContentListNode(nodes).toDiagnosticsNode(name: 'item $i'))
      .toList();
  }
```

PIG208 commented:
Update: resolve a merge conflict and make minor formatting change.

gnprice reviewed (commented):
Thanks @PIG208, and thanks @chrisbobbe for the previous review!

Comments below.

Also one higher-level comment: the PollWidget depends on the mutable data in Poll, which isn't how Flutter expects things to work: see docs on [StatelessWidget.build]. Instead let's make Poll into a ChangeNotifier, and PollWidget into a stateful widget that listens to it.

That will also let us skip notifying the entire MessageListView's listeners when a poll gets updated, which will be good for performance in some scenarios: if you're looking at a thread with a poll that's actively getting votes, it'll let us avoid rebuilding all the other messages every time there's a vote on the poll.

(I think this version ends up behaving correctly because the PollWidget gets rebuilt as a descendant of the MessageList, but best to align with the architecture anyway. That aligning with the architecture is good for performance is a bonus.)

PIG208 commented on lib/widgets/content.dart:
```diff
@@ -222,17 +251,20 @@ const double kBaseFontSize = 17;
 /// or the message's status as starred or edited.
 class MessageContent extends StatelessWidget {
-  const MessageContent({super.key, required this.message, required this.content});
+  const MessageContent({super.key, required this.message, required ZulipMessageContent content}) : _content = content;
```
This is for type promotion, similar to [this other instance](https://github.com/zulip/zulip-flutter/pull/808#issue-2406389310).

gnprice replied:
> narrow is refactored to become private because otherwise `switch` is not able to promote the type of it.

I see. In that case:
https://github.com/zulip/zulip-flutter/pull/808#discussion_r1677245004

:slightly_smiling_face: 

In general, there's a local fix for that problem at any place where it arises, whereas changing a field from public to private changes the class's API for everywhere. So it's better to not let this affect whether the field is public or private.

PIG208 commented:
Updated the PR. Now we are using `localizedTextBaseline` to align the text. This makes a slight difference:

| before | after |
| - | - |
| ![Screenshot from 2024-09-18 21-46-28](https://github.com/user-attachments/assets/207e4bff-6875-4ac2-965b-1769a39eb779) | ![Screenshot from 2024-09-18 21-45-32](https://github.com/user-attachments/assets/b4726b83-5a10-4fbe-be06-176023b3e329) |

Made `Poll` a `ChangeNotifier`, and let `Poll` listen to `Poll` instead. The second to last commit summary is intended to be discarded as the final commit squashes into it.

gnprice commented on test/model/message_list_test.dart:
```diff
@@ -544,65 +509,76 @@ void main() {
-      await prepareMessages(foundOldest: false,
-        messages: messages.where((m) => m.id != 151).toList());
-      model.notifyListenersIfAnyMessagePresent([150, 151, 152]);
+    test('parse PollContent', () async {
```
These two new cases don't belong in a group that's about `handleMessageEvent`, because they're not really about that method — in particular they're not any more specific to that method than they are to `fetchInitial` or `fetchOlder`.

In general this test file is one where our usual principle of organizing the tests in the same order as the code they test often breaks down: `model/message_list.dart` is one of our most intricate areas of model code, and might be the one file with the greatest complexity of interrelationships between different parts of its implementation.

You can see that organizing principle already broken down in roughly the latter half of the file: the "regression tests for …" and "stream/topic muting" groups, the tests "showSender is maintained correctly" and "recipient headers are maintained consistently", and the `checkInvariants` function which underlies most of the tests and describes invariants that cut across the implementation code.

So these can go in a new small group of their own, perhaps just above "recipient headers are maintained consistently" since they're both about logic within _MessageSequence.

Then I think we don't need these commits:
1919c1ba2 msglist test [nfc]: Make a group for MessageEvent tests
99a62b71a msglist test [nfc]: Name test groups after the event handler method
b69f0bfef msglist test [nfc]: Move notifyListeners tests to match implementation

gnprice commented on test/model/message_list_test.dart:
```diff
@@ -1780,32 +1824,6 @@ void checkInvariants(MessageListView model) {
-  Subject<bool> get isLastInBlock => has((x) => x.isLastInBlock, 'isLastInBlock');
-}
-
-extension MessageListViewChecks on Subject<MessageListView> {
```
So, the downside of moving these checks-extensions to their own file is that its name `message_list_checks.dart` starts the same way as the test file `message_list_test.dart`. That makes it get in the way of tab-completion, when writing the test file's name for a `git log` or `flutter test` command.

For that reason I've been keeping the extensions within the test file when they're pretty specific to the one file, or to that and one or two others. I think these are best kept here for that reason.

gnprice reviewed (commented):
Thanks for the revision! Generally this looks good — a few comments below.

I haven't read closely all the details of the build method, but I'll skip doing so; this seems good enough.

Though it looks like @alya hasn't been consulted on this thread yet. Alya, any feedback on the UI in the screenshots above?

> The second to last commit summary is intended to be discarded as the final commit squashes into it.

I recommend making the commit messages into a draft of what you'd like to see merged. :slightly_smiling_face: This sort of remark would be easy for me to miss.

In general if there are details you want to say only at review time and don't think belong in the Git history, the PR thread (or a chat thread) is a good venue for those.

In that commit message I think the details are good for the Git history, though:

    The UI follows the webapp until we get a new design.
    
    The dark theme colors were tentatively picked. The `TextStyle`s are
    the same for both light and dark theme. All the styling are based on
    values taken from the webapp.
    
    References:
    
      - light theme:
    
        https://github.com/zulip/zulip/blob/2011e0df760cea52c31914e7b77d9b4e38e9ee74/web/styles/widgets.css#L138-L185
        https://github.com/zulip/zulip/blob/2011e0df760cea52c31914e7b77d9b4e38e9ee74/web/styles/dark_theme.css#L358
    
      - dark theme:
    
        https://github.com/zulip/zulip/blob/2011e0df760cea52c31914e7b77d9b4e38e9ee74/web/styles/dark_theme.css#L966-L987
    
They give good context for what we were thinking when we made the particular decisions that appear in that code. That's the sort of thing that's helpful to a later reader who's, for example, thinking about changing the code and wants to know what background they might be missing.

alya commented:
Looks reasonable to me! 

I remember making some comments about polls earlier (which I believe have been addressed), but I guess that was on a different PR.

PIG208 commented:
The second to last commit is to be kept now, because previous those two were in the reverse order because of the way the tests were organized. My (stale) comment on that did make it more confusing!

Pushed an update addressing the review comments.

gnprice commented:
Thanks for the revision! This all looks good — merging. Happy to get this feature in.

I took out one new paragraph the latest revision added to a commit message:

    -    _reparseContent currently happens when there is content edit, which is
    -    forbidden for poll messages, (and you can't turn an existing message
    -    into a poll).  In most cases the same `PollContent` instance will be
    -    used for each poll message, unlike `ZulipContent`, and a poll message
    -    never needs to get reparsed.

It's true that over the lifetime of a given MessageListView, in a release build, a given poll message will be represented by one PollContent instance. But it can get a new PollContent if `MessageListView.reassemble` is called, which happens routinely in development. And in all builds, if the same message appears in a new message list, it'll get a new PollContent instance there. So there's no guarantee of uniqueness.

From a performance perspective, the main point here is that allocating one of these trivial `PollContent` wrapper objects is a much cheaper operation than parsing a message's HTML, which is what we do for normal messages in all the same situations where we make a `PollContent` for polls.
